frame/dvm/src/lib.rs

```diff
@@ -332,37 +349,154 @@ pub mod pallet {
+	fn validate_transaction_common(
+		origin: H160,
+		transaction: &TransactionV0,
+	) -> Result<U256, TransactionValidityError> {
+		// We must ensure a transaction can pay the cost of its data bytes.
+		// If it can't it should not be included in a block.
+		let mut gasometer = evm::gasometer::Gasometer::new(
+			transaction.gas_limit.low_u64(),
+			<T as darwinia_evm::Config>::config(),
 		);
-		let transaction = Self::to_dvm_transaction(transaction)?;
-		Self::raw_transact(transaction).map(|(_, used_gas)| {
-			Ok(Some(T::GasWeightMapping::gas_to_weight(
-				used_gas.unique_saturated_into(),
+		let transaction_cost = match transaction.action {
+			TransactionAction::Call(_) => evm::gasometer::call_transaction_cost(&transaction.input),
+			TransactionAction::Create => {
+				evm::gasometer::create_transaction_cost(&transaction.input)
+			}
+		};
+		if gasometer.record_transaction(transaction_cost).is_err() {
+			return Err(InvalidTransaction::Custom(
+				TransactionValidationError::InvalidGasLimit as u8,
+			)
+			.into());
+		}
+
+		if let Some(chain_id) = transaction.signature.chain_id() {
+			if chain_id != T::ChainId::get() {
+				return Err(InvalidTransaction::Custom(
+					TransactionValidationError::InvalidChainId as u8,
+				)
+				.into());
+			}
+		}
+
+		if transaction.gas_limit >= T::BlockGasLimit::get() {
+			return Err(InvalidTransaction::Custom(
+				TransactionValidationError::InvalidGasLimit as u8,
+			)
+			.into());
+		}
+
+		let account_data = <T as darwinia_evm::Config>::RingAccountBasic::account_basic(&origin);
+		let fee = transaction.gas_price.saturating_mul(transaction.gas_limit);
+		let total_payment = transaction.value.saturating_add(fee);
+		if account_data.balance < total_payment {
+			return Err(InvalidTransaction::Payment.into());
+		}
+
+		let min_gas_price = T::FeeCalculator::min_gas_price();
+		if transaction.gas_price < min_gas_price {
+			return Err(InvalidTransaction::Payment.into());
+		}
+
+		Ok(account_data.nonce)
+	}
+
+	// Controls that must be performed by the pool.
+	// The controls common with the State Transition Function (STF) are in
+	// the function `validate_transaction_common`.
+	fn validate_transaction_in_pool(
+		origin: H160,
+		transaction: &TransactionV0,
+	) -> TransactionValidity {
+		let account_nonce = Self::validate_transaction_common(origin, transaction)?;
+		if transaction.nonce < account_nonce {
+			return Err(InvalidTransaction::Stale.into());
+		}
+
+		// The tag provides and requires must be filled correctly according to the nonce.
+		let mut builder = ValidTransactionBuilder::default()
+			.and_provides((origin, transaction.nonce))
+			.priority(transaction.gas_price.unique_saturated_into());
+
+		// In the context of the pool, a transaction with
+		// too high a nonce is still considered valid
+		if transaction.nonce > account_nonce {
+			if let Some(prev_nonce) = transaction.nonce.checked_sub(1.into()) {
+				builder = builder.and_requires((origin, prev_nonce))
+			}
+		}
```

Comment:
do we need to limit this nonce?
If users send so many txs that nonce very large. Whether the transaction pool will drop it?

Reply:
No, we don't need to do this. Here we use substrate tx pool, the limit task should be controlled there.

Reply:
If I attack the tx pool by sending so many txs with invalid nonce(nonce > currentNonce) continuously. How can u process the pool？

Reply:
1. If all the txs with the same nonce(large than the current one), will trigger the `TemporarilyBanned` mechanism, then you have to stop your attack behavior.
2. If all the txs with different nonce(increase one by one), there is a max limit for transaction future pool, If reached, then trigger enforce clean up. 
- https://github.com/paritytech/substrate/blob/master/client/transaction-pool/src/graph/validated_pool.rs#L239
- https://github.com/paritytech/substrate/blob/master/client/transaction-pool/src/graph/pool.rs#L113